### PR TITLE
Implement -I and -J (PassOptions edition)

### PR DESCRIPTION
**Pull request by meow464**
[Old](https://github.com/lfortran/lfortran/pull/901)

This implements -I and -J using the PassOptions strtuct.

`lfortran -c sum.f90 -J mods/` 
Will generate mods/my_module.mod

`lfortran example.f90 -J mods/` (-I works here too)
Will compile, run and output 18.000000

example.f90
```
program example
  use my_module
  implicit none
  real :: x,y

  x = 8.0
  call sum_ten(x,y)
  print *, y
end program example
```

sum.f90
```
module my_module
  contains
    subroutine sum_ten(a,b)
      implicit none
      real, intent(in) :: a
      real, intent(out) :: b
      b = a + 10
    end subroutine sum_ten
end module my_module
```

### CODE REVIEW

**Comment by meow464** [comment]
@certik Ready for review.

One note: Currently you have to manually call `pass_options.runtime_library_dir = LFortran::get_runtime_library_dir()`. This could be put on the PassOptions constructor but:

1. PassOptions is declared in libasr/utils.h
2. get_runtime_library_dir() is declared in lfrotran/utils.h, which includes libasr/utils.h , so we would have a circular include

Solutions are:
1. Forwards declare `get_runtime_library_dir()` in `libasr/utils.h`. I'm not  fan of forward declares and from what I understand libasr should be an independent library.
2. Move `get_runtime_library_dir()` to `libasr/utils.h`, this requires also moving `whereami.h` to libasr. You could make a case for `get_runtime_library_dir()` not belonging in the asr library.

**Comment by certik** [comment]
Awesome, I think it looks much cleaner. Let me try to review tomorrow. We need this. Thanks!

**Comment by meow464** [comment]
Some more context: It makes sense to start with PassMnager.apply_passes(), it's the choke point. It calls get_fma, get_flipsign and get_sign_from value, which indirectly call find_and_load_module() and need to send PassOptions down to it. Everything else indirectly called PassManager.apply_passes() and thus needed to send PassOptions down to it.

**Comment by certik** [comment]
I am reviewing this now.

CC @czgdp1807 for the pass infrastructure feedback. We need to iteratively keep improving it.

**Comment by certik** [comment]
Ok, if I merge https://github.com/lfortran/lfortran/pull/945 into this, then it compiles Minpack correctly and no need for manual moving of `.mod` files!

**Comment by certik** [comment]
I went over the patch, it seems it is now using PassOptions instead of CompilerOptions. It adds a lot of maintenance into `lfortran.cpp`, I don't think we should be passing it like this there.

I am assigning this PR to myself, let me try to figure out a solution that is simpler.

**Comment by meow464** [comment]
Do you think it would be better to put mod_files_dir and include_dirs on compiler_options and only set it on pass_options at the libasr boundary (apply_passes)? Most functions on lfortran.cpp already receive compiler_options anyways. This would make so FortranEvaluator can create its own pass_options instance and, I think, wouldn't require the test changes either.

**Comment by meow464** [comment]
I could do that today.

**Comment by meow464** [comment]
This would allow us to use get_runtime_library_dir() on the CompilerOptions constructor and thus avoid a lot of random calls through the code.

**Comment by certik** [comment]
Yes, something like that is what I was thinking. Essentially minimize the changes needed to implement this.

**Comment by meow464** [comment]
@certik It's passing.

I could:

1. Forward declare get_runtime_library_dir() to use on the CompilerOptions constructor and
2. Create a new struct containing mod_files_dir, include_dirs and runtime_library_dir and make it a member of both CompilerOptions  and PassOptions. This would make passing the options just pass_options.dirs = compiler_options.dirs.

These things could simplify or complicate the PR, depending on who you ask.

File: src/lfortran/fortran_evaluator.cpp [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #include <lfortran/parser/parser.h>
 #include <lfortran/parser/preprocessor.h>
 #include <lfortran/pickle.h>
+#include <libasr/utils.h>
```
Comment (certik) [inline src/lfortran/fortran_evaluator.cpp:18]: This is not needed anymore.

File: src/lfortran/fortran_evaluator.h [hunk 2]
```diff
@@ -13,6 +13,7 @@
 #include <libasr/config.h>
 #include <libasr/diagnostics.h>
 #include <libasr/pass/pass_manager.h>
+#include <libasr/utils.h>
```
Comment (certik) [inline src/lfortran/fortran_evaluator.h:16]: probably not needed

**Review by certik — APPROVED** [review]
I think this is good enough. We probably can, as you said, not pass PassOptions into load_module(), but instead probably pass CompilerOptions. That is a simple change that can be done in a future PR. I am going to squash and merge this.

Thanks for your hard work on this and I am sorry it took so many iterations.

**Comment by certik** [comment]
I tested one more time and it compiles minpack without any hand moving files:
```console
$ make clean
$ make
[  2%] Building Fortran object src/CMakeFiles/minpack.dir/minpack.f90.o
[  4%] Building Fortran object src/CMakeFiles/minpack.dir/covar.f.o
[  7%] Building Fortran object src/CMakeFiles/minpack.dir/errjac.f.o
[  9%] Building Fortran object src/CMakeFiles/minpack.dir/hybipt.f.o
[ 12%] Building Fortran object src/CMakeFiles/minpack.dir/lhesfcn.f.o
[ 14%] Building Fortran object src/CMakeFiles/minpack.dir/lmdipt.f.o
[ 17%] Building Fortran object src/CMakeFiles/minpack.dir/ocpipt.f.o
[ 19%] Building Fortran object src/CMakeFiles/minpack.dir/r1updt.f.o
[ 21%] Building Fortran object src/CMakeFiles/minpack.dir/vecjac.f.o
[ 24%] Building Fortran object src/CMakeFiles/minpack.dir/chkder.f.o
[ 26%] Building Fortran object src/CMakeFiles/minpack.dir/dmchar.f.o
[ 29%] Building Fortran object src/CMakeFiles/minpack.dir/fdjac1.f.o
[ 31%] Building Fortran object src/CMakeFiles/minpack.dir/hybrd1.f.o
[ 34%] Building Fortran object src/CMakeFiles/minpack.dir/lmder1.f.o
[ 36%] Building Fortran object src/CMakeFiles/minpack.dir/lmpar.f.o
[ 39%] Building Fortran object src/CMakeFiles/minpack.dir/qform.f.o
[ 41%] Building Fortran object src/CMakeFiles/minpack.dir/rwupdt.f.o
[ 43%] Building Fortran object src/CMakeFiles/minpack.dir/dogleg.f.o
[ 46%] Building Fortran object src/CMakeFiles/minpack.dir/fdjac2.f.o
[ 48%] Building Fortran object src/CMakeFiles/minpack.dir/hybrd.f.o
[ 51%] Building Fortran object src/CMakeFiles/minpack.dir/lmder.f.o
[ 53%] Building Fortran object src/CMakeFiles/minpack.dir/lmstr1.f.o
[ 56%] Building Fortran object src/CMakeFiles/minpack.dir/qrfac.f.o
[ 58%] Building Fortran object src/CMakeFiles/minpack.dir/ssqfcn.f.o
[ 60%] Building Fortran object src/CMakeFiles/minpack.dir/dpmpar.f.o
[ 63%] Building Fortran object src/CMakeFiles/minpack.dir/grdfcn.f.o
[ 65%] Building Fortran object src/CMakeFiles/minpack.dir/hybrj1.f.o
[ 68%] Building Fortran object src/CMakeFiles/minpack.dir/lmdif1.f.o
[ 70%] Building Fortran object src/CMakeFiles/minpack.dir/lmstr.f.o
[ 73%] Building Fortran object src/CMakeFiles/minpack.dir/qrsolv.f.o
[ 75%] Building Fortran object src/CMakeFiles/minpack.dir/ssqjac.f.o
[ 78%] Building Fortran object src/CMakeFiles/minpack.dir/enorm.f.o
[ 80%] Building Fortran object src/CMakeFiles/minpack.dir/hesfcn.f.o
[ 82%] Building Fortran object src/CMakeFiles/minpack.dir/hybrj.f.o
[ 85%] Building Fortran object src/CMakeFiles/minpack.dir/lmdif.f.o
[ 87%] Building Fortran object src/CMakeFiles/minpack.dir/objfcn.f.o
[ 90%] Building Fortran object src/CMakeFiles/minpack.dir/r1mpyq.f.o
[ 92%] Building Fortran object src/CMakeFiles/minpack.dir/vecfcn.f.o
[ 95%] Linking Fortran static library libminpack.a
[ 95%] Built target minpack
[ 97%] Building Fortran object examples/CMakeFiles/example_hybrd.dir/example_hybrd.f90.o
warning: Format statement is not implemented yet, for now we will ignore it
  --> /Users/ondrej/repos/lfortran/minpack/examples/example_hybrd.f90:54:6 - 57:58
   |
54 |    1000 format(5x, "FINAL L2 NORM OF THE RESIDUALS", d15.7// &
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
...
   |
57 |               5x, "FINAL APPROXIMATE SOLUTION"//(5x, 3d15.7))
   | ...^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ignored for now


Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that needs to be fixed).
warning: unit in write() is not implemented yet and it is currently treated as '*'
  --> /Users/ondrej/repos/lfortran/minpack/examples/example_hybrd.f90:52:12
   |
52 |     write (nwrite, 1000) fnorm, nfev, info !, (x(j), j=1, n)
   |            ^^^^^^ treated as '*'

warning: format string in write() is not implemented yet and it is currently treated as '*'
  --> /Users/ondrej/repos/lfortran/minpack/examples/example_hybrd.f90:52:20
   |
52 |     write (nwrite, 1000) fnorm, nfev, info !, (x(j), j=1, n)
   |                    ^^^^ treated as '*'


Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that needs to be fixed).
[100%] Linking Fortran executable example_hybrd
[100%] Built target example_hybrd
```

**Pull request merged**